Make Game Code as Class

raymiS opened the pull request:
**Status**

Ready

**Implementation Notes**

Finishing the basic game functionality I decided to improve it by creating a class that will receive the word that the player needs to guess.

raymiS committed:
Change array method in word_spaces_empty from all? to any?

This will help the word_spaces_empty? to return true even if one of the spaces in the array is equal to ""

raymiS committed:
Make Game Code Structure as a Class

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,44 @@
-def display_game
-  @word_spaces.each do |characters|
-    print "| #{characters} | "
+class Ahorcado
```
I'm fine with using Spanish names for titles and stuff. However, have you thought about looking at the English name of this game?

raymiS replied:
Yes, I will update that 💃

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,44 @@
-    print "| #{characters} | "
+class Ahorcado
+  def initialize(current_word)
+    @current_word = current_word
```
To make sure, you'll be creating a new "Ahorcado" Object per each game instance, right?

raymiS replied:
Yes, since this is a one-player game, a new game should be created again once the player either wins or loses.

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,44 @@
+      print "| #{characters} | "
+    end
+    choose_character
   end
```
Is it important to `choose_character` every time you `display_game`? Seems like they're related but wouldn't necessarily be chained together.

raymiS replied:
I will look into that.

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,44 @@
-end
+  def word_spaces_empty?
+    @word_spaces.any? { |w| w == "" }
+  end
```
Can you think of other ways to handle evaluating empty spaces?

raymiS replied:
Yes, I just push an update. Is that the change you were referring to?

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,44 @@
-    puts "Guess the word! Choose a letter"
-    chosen_character = gets.chomp
+  def choose_character
+    while word_spaces_empty? do
```
The naming here sounds weird. Could you think in a better name that sounds better?

` while [SOMETHING IS HAPPENING] do`

Don't worry about being a bit verbose. Most of the times, that's welcomed.

raymiS replied:
Sure, I will change it 💃

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,44 @@
+      insert_character(chosen_character)
 
-    insert_character(chosen_character)
+      puts "#{word_spaces_empty?}"
```
What's the reason for this line?

raymiS replied:
Oh it was for debbuging. I'll remove it.

alessandrapereyra reviewed (changes requested):
Looks good so far. I've added some comments and RFC.

raymiS committed:
Change name of game for the English equivalent.

raymiS committed:
More Verbose Method Name

  Change word_spaces_empty? to is_character_space_empty? method's name for better understanding of what the method is evaluating.

raymiS committed:
Remove a puts that was there for debugging reasons.

raymiS committed:
Move Method for Choosing a Character

This method doesn't need to be executed every time de game is displayed. Reason to why I decided to put it after the game initialize.

raymiS committed:
Use include?

  Change the way that I evaluate if there are empty spaces on the @word_spaces array. Instead of any? I use include?

raymiS committed:
Add method for testing winning and losing scenarios.

  Add method has_player_lost to check whether or not the player has lost. This will happen when the number of chances a player has is equal to the length of the word they have to guess ( @current_word )

Also add a winner message once the player has filled all the word_spaces with the correct characters.

raymiS commented:
@alessandrapereyra: I've made a few changes and fix the ones that were requested. Whenever you have a spare time, let me know if there's anything else you'd like me to do 💃

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,57 @@
-  @word_spaces.each do |characters|
-    print "| #{characters} | "
+class Hangman
+  def initialize(current_word)
```
Since this argument is seeding the new data, could you think of a better word for this one? 

`@current_word` kinda works well internally but makes me think there could be a moment when there's another word. Since right now you're tying down one word per instance, not sure it represents what's actually happening.

raymiS replied:
Changed it for `@game_word` 💃

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,57 @@
+    end
+  end
+
+  def is_character_space_empty?
```
Not sure what `character_space_empty` means. The code inside relates to `@word_spaces` having blank/empty spaces, but not sure how that relates to a `character_space` or whether it'd be empty.

raymiS replied:
I tried with something longer and more specific like `is_word_spaces_arr_empty?` 😄

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,57 @@
+    @current_word = current_word
+    @word_spaces = Array.new(current_word.length, "")
+    @current_characters = current_word.split("")
+    @loser_count = 0
```
This leads me to think you're counting whenever someone loses the game (which signifies becoming a loser). Right now, though, seems like it's really depicting the failing attempts made (since the max number of them you're comparing is the `@current_word.length`. 

Could you think of a better wording?

raymiS replied:
Since I'm counting the chances the player has to guess the word I decided to use `@player_guess_chances`

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,57 @@
-  while word_spaces_empty? do
-    puts "Guess the word! Choose a letter"
-    chosen_character = gets.chomp
+  def has_player_lost
```
Usually, the naming `has_x` naming scheme is used to evaluate the existence of something, and return `true` or `false` if t does (or not). In this case, you're correctly validating someone has lost, but also taken other actions (like rendering something to the screen). That one should probably be used elsewhere as to reduce the responsibilities this method has to handle.

raymiS replied:
Fixed.

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,57 @@
+  end
+
+  def choose_character
+    while is_character_space_empty? && !has_player_lost do
```
Sometimes adding methods that invalidate another are a good trick to skip using "!" and such on conditions: 

```
def player_still_on_game?
  !has_player_lost
end
```

Thus, your condition can be read as

`while is_character_space_empty? && player_still_on_game? do`

raymiS replied:
Thank you for the tip. Definitely, used it 😄

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,57 @@
+    end
 
-    insert_character(chosen_character)
+    if !is_character_space_empty?
```
Try to think of ways  to express this without negatives (either with a helper method as above or with another condition)

raymiS replied:
In this case, since I change `has_player_lost` method to only be boolean I decided to put both screen messages on the same conditional so I didn't nee this line any longer 😸

alessandrapereyra commented on ahorcado.rb:
```diff
@@ -1,62 +1,57 @@
-    puts "Word Spaces: #{@word_spaces}"
-  else 
-    puts "Nope, keep playin"
+    has_player_lost
```
Seems like you'd be coupling the act of adding chosen characters to the array with an evaluation if that let them win.

raymiS replied:
Fixed.

raymiS committed:
Start Game Method

  The choose_character method represented the start of the game right after the character spaces are displayed.

raymiS committed:
Remove Extra Responsability for insert_character method

  The insert_character responsability is to place the character inside @word_spaces correct position if that character is part of the word theplayer is trying to guess.

  This fixes the previous extra work that this method had, which was to also evaluate what to do in case the character is NOT part of thewordthe player is trying to guess. This logic will go inside the start_game method.

raymiS committed:
Second try at naming is_character_space_empty? method

  In this second try I base the name of the method more verbose and specific for better understanding of what the method is checking

raymiS committed:
Remove screen message from has_player_lost method

This method is a boolean and it's only responsability is to return true or false if the player has run out of guessing chances.

The screen message for win or lose logic will be at the start_game method. This way I can have all the game logic in one place.

raymiS committed:
Change current_word variable and argument name

Since this game starts for every instance I only need one word: game_word

raymiS committed:
Change loser_count variable name

For player_guess_chances for better understanding of what the variable is keeping count

raymiS committed:
Improve Display Messages

  Work on the display messages for better game flow

raymiS committed:
Clean Rebase Code
